soft4pes/control/mpc/solvers/mpc_bnb.py

```diff
@@ -0,0 +1,129 @@
+
+        # Create all possible three-phase switch positions
+        self.SW_COMB = np.array(list(product(sw_pos_3ph, repeat=3)))
+
+    def __call__(self, sys, conv, ctr, y_ref):
+        """
+        Solve MPC problem by using a simple BnB method.
+
+        Parameters
+        ----------
+        sys : system object
+            System model.
+        conv : converter object
+            Converter model.
+        ctr : controller object
+            Controller model.
+        y_ref : 1 x _ ndarray of floats
+            Reference vector [p.u.].
+
+        Returns
+        -------
+        uk : 1 x 3 ndarray of ints
+            The three-phase switch position.
+        """
+
+        self.J_min = np.inf
+        J_prev = 0
+        self.U_seq = np.zeros(3 * ctr.Np)
+        self.U_temp = np.zeros(3 * ctr.Np)
+        k = 0
+
+        self.solve(sys, conv, ctr, sys.x, y_ref, ctr.u_km1, k, J_prev)
+
+        uk = self.U_seq[0:3]
+        return uk
+
+    def solve(self, sys, conv, ctr, xk, y_ref, u_km1, k, J_prev):
+        """
+        Recursively compute the cost for different switching sequences.
+
+        Parameters
+        ----------
+        sys : object
+            System model.
+        conv : object
+            Converter model.
+        ctr : object
+            Controller model.
+        xk : 1 x _ ndarray
+            Current state vector [p.u.].
+        y_ref : 1 x _ ndarray
+            Reference vector [p.u.].
+        u_km1 : ndarray
+            Previous three-phase switch position.
+        k : int
+            Prediction step.
+        J_prev : float
+            Previous cost.
+        """
+
+        # Iterate over all possible three-phase switch positions
+        for uk in self.SW_COMB:
+
+            # Check if switching constraint is violated or cost is infinite
+            if not switching_constraint_violated(conv.nl, uk, u_km1):
+
+                # Compute the next state
+                x_kp1 = ctr.get_next_state(sys, xk, uk, k)
+
+                # Calculate the cost of reference tracking and control effort
+                y_error = np.linalg.norm(y_ref[k] - np.dot(ctr.C, x_kp1))**2
+                delta_u = ctr.lambda_u * np.linalg.norm(uk - u_km1, ord=1)
+                J_temp = J_prev + y_error + delta_u
+
+                # if the cost is smaller than the current minimum cost, continue
+                if J_temp < self.J_min:
+
+                    # If not at the last prediction step, move to the next prediction step
+                    if k < ctr.Np - 1:
+                        self.u_temp[3 * k:3 * (k + 1)] = uk
+                        self.solve(sys, conv, ctr, x_kp1, y_ref, uk, k + 1,
+                                   J_temp)
+                    else:
+                        # If at the last prediction step, store the three-phase switch position and
+                        # update the minimum cost
+                        self.U_temp[3 * k:3 * (k + 1)] = uk
```

Comment:
Is this u(k) or u(k+Np-1)?

Reply:
I think this relates to the comment I made above about which part of the sequence is considered as the applied three-phase switch position.

As for the naming, I think this needs to be fixed as soon as possible. At the moment, some definitions look confusing (or contradictory) and it seems that some comments will just re-appear as long as this issue has not been addressed